pep-0646.rst

```diff
@@ -743,11 +753,135 @@ is available in `cpython/23527`_. A preliminary version of the version
+  *only* be willing to take array arguments that have image-like shapes, this cannot be stipulated
+  with named tensors.
+* **Static checking** of shape correctness is still not possible.
+* **Poor uptake**. Because each library much implement its own form of named tensors,
+  the burden on library maintainers in onerous - and as a result, named tensors have not
+  seen as widespread support as we might have hoped for.
+  
+Can the 'named tensors' approach be combined with the approach we advocate for in
+this PEP? We're not sure yet. One area of overlap is that in some contexts, we could do, say,
+``image: Array[Height, Width, Channels]; mean(image, axis=Channels)``. Ideally,
+we might write something like ``image: Array[Height=64, Width=64, Channels=3]`` -
+but this won't be possible in the short term, due to the rejection of PEP 637.
+In any case, our attitude towards this is mostly "Wait and see what happens before
+taking any further steps".
+
+Named Axes vs Literal Shape Specification
+-----------------------------------------
+
+As we mentioned in `Summary Examples`_, named axes are not the only use of this PEP.
+Instead of using axis names to parameterise array types, we could also parameterise
+with the actual sizes, e.g. ``Array[Literal[64], Literal[64], Literal[3]]``. This approach
+more naturally fits with the idea of shape arithmetic - perhaps in combination
+with something like a type variable-like object but for integers, such that we could
+write ``Array[N] -> Array[2*N]``.
+
+As of writing, we're genuinely unsure which approach will be most fruitful long-term.
+Here too, out attitude is "Take only the steps which we're confident are universally
+beneficial (this PEP) and wait and see what happens from there before committing to
+a specific path."
+
+Meaning and Scope of Axis Names
+-------------------------------
+
+If we attach a name like ``Batch`` to a particular axis, what does that actually mean?
+Is that name a placeholder for some actual value, like 32 - such that we're stipulating
+that all arrays with an axis named `Batch` must have the same batch size?
+
+And if so, what should the scope of such labels be? The local function? The module?
+The whole program?
+
+Or is the name ``Batch`` merely referring to the identity, the semantic meaning, of
+the axis? Under this interpretation, different arrays with an axis named ``Batch`` could
+have different sizes, and ``Batch`` would only serve as indication of which axis
+served as the batch-like axis.
+
+And if so, again, what should the scope of such labels be? Should we consider all axes labelled
+as 'Batch' to have the same identity, even if in completely different modules?
+
+We espouse the second view of axis names - names as indicative of semantic identity,
+but *not* of actual size. To put it another way, we think that axis names should behave
+like variable *types*, not variable *names*:
+
```

Comment:
I agree that this is a useful definition (especially for application code), but I think it's also worth explicitly describing how typing for fixed sizes would fit into this model, because there are very clear use cases for it in library code.

For example, libraries like NumPy would want to be able to write something like:
```python
def matrix_matrix_muliply(x: Array[K, N], Array[N, M]) -> Array[K, M]: ...
def matrix_vector_multiply(x: Array[K, N], Array[N]) -> Array[K]: ...
```
where named axes on the _same_ function indicate a shared size, but there is no implied relationship between the axis names `K` and `N on different functions.

(Note: from our offline discussions, I'm pretty sure this is supported with type variables, but I'd still like to see it in the PEP for the sake of clarification for maintainers of array libraries.)

Reply:
Ahh, right, this brings back to mind a whole mess of context I'd forgotten to write down. I've expanded this section - what do you think of it now?

Reply:
(I've talked to Stephan offline and I'm pretty sure he's happy with this now.)